WordPressAuthenticatorTests/Authenticator/WordPressAuthenticatorTests.swift

```diff
@@ -38,21 +45,352 @@ class WordPressAuthenticatorTests: XCTestCase {
+        XCTAssertEqual(controller.loginFields.meta.jetpackBlogUsername, "username")
+        XCTAssertEqual(controller.loginFields.username, "email-address@example.com")
+    }
+    
+    func testShowLoginForSelfHostedSitePresentsCorrectVC() {
+        let presenterSpy = ModalViewControllerPresentingSpy()
+        let expectation = XCTNSPredicateExpectation(predicate: NSPredicate(block: { (_, _) -> Bool in
+            return presenterSpy.presentedVC != nil
+        }), object: .none)
+        
+        WordPressAuthenticator.showLoginForSelfHostedSite(presenterSpy)
+        wait(for: [expectation], timeout: 3)
+        
+        XCTAssertTrue(presenterSpy.presentedVC is LoginNavigationController)
+    }
+    
+    func testShowLoginForSelfHostedSiteTracksOpenLogin() {
+        let presenterSpy = ModalViewControllerPresentingSpy()
+        let delegateSpy = WordPressAuthenticatorDelegateSpy()
+        
+        WordPressAuthenticator.shared.delegate = delegateSpy
+        
+        let expectation = XCTNSPredicateExpectation(predicate: NSPredicate(block: { (_, _) -> Bool in
+            return presenterSpy.presentedVC != nil
+        }), object: .none)
+        
+        WordPressAuthenticator.showLoginForSelfHostedSite(presenterSpy)
+        wait(for: [expectation], timeout: 3)
+        
+        let trackedEvent = delegateSpy.trackedElement
+        
+        XCTAssertEqual(trackedEvent, WPAnalyticsStat.openedLogin)
+    }
+    
+    func testSignInForWPComReturnsVC() {
+        let vc = WordPressAuthenticator.signinForWPCom()
+        
+        XCTAssertTrue((vc as Any) is LoginEmailViewController)
+    }
+    
+    func testSignInForWPComWithLoginFieldsReturnsVC() {
+        let navController = WordPressAuthenticator.signinForWPCom(dotcomEmailAddress: "example@email.com", dotcomUsername: "username") as! UINavigationController
+        let vc = navController.topViewController
+        
+        XCTAssertTrue((navController as Any) is UIViewController)
+        XCTAssertTrue((vc as Any) is LoginWPComViewController)
+    }
+    
+    func testSignInForWPComSetsEmptyLoginFields() {
+        let navController = WordPressAuthenticator.signinForWPCom(dotcomEmailAddress: nil, dotcomUsername: nil) as! UINavigationController
+        let vc = navController.topViewController as! LoginWPComViewController
+        
+        XCTAssertEqual(vc.loginFields.emailAddress, String())
+        XCTAssertEqual(vc.loginFields.username, String())
+    }
+    
+    func testTrackOpenedLoginSendsCorrectTrackValue() {
+        let delegate = WordPressAuthenticatorDelegateSpy()
+        WordPressAuthenticator.shared.delegate = delegate
+        
+        WordPressAuthenticator.trackOpenedLogin()
+        
+        guard let trackedEvent = delegate.trackedElement else {
+            XCTFail("Event not Tracked")
+            return
+        }
+        
+        XCTAssertEqual(trackedEvent, WPAnalyticsStat.openedLogin)
```

Comment:
If you were to do the following, the test would still fail in case the event wasn't tracked and you'd have a more compact test.

```suggestion
        XCTAssertEqual(delegate.trackedElement, WPAnalyticsStat.openedLogin)
```

What do you think?

P.S. I didn't try compiling this code. If it doesn't compile, please ignore my comment.

Reply:
Totally, much clearer!